### PR TITLE
Move logic for quorum check to the backend

### PR DESCRIPTION
**Pull request by vcastellm**
# Description

First step before implementing voting power is to move the current quorum check to the backend. We're maintaining here the current quorum messages counting and the implementation on the backend will maintain the current logic, so nothing changes in this PR.

See https://www.notion.so/polygontechnology/RFC-21-f2fd4a77333c494cbb5b96f0054bd5ef

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [x] New feature (non-breaking change that adds functionality)
- [x] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Breaking changes

There's a new method that the backend using this library should implement `HasQuorum`

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the relevant labels
- [x] I have added sufficient documentation in code

### CODE REVIEW

File: core/mock_test.go [hunk 1]
```diff
@@ -181,6 +181,21 @@ func (m mockBackend) BuildRoundChangeMessage(
...
 }
 
+func (m mockBackend) HasQuorum(blockNumber uint64, messages []*proto.Message) bool {
+	quorum := m.Quorum(blockNumber)
+
+	if len(messages) > 0 {
+		switch messages[0].GetType() {
+		case proto.MessageType_PREPARE:
+			return len(messages) < int(quorum)-1
+		case proto.MessageType_ROUND_CHANGE:
```
[the remark this answers is not in the document]
Comment (vcastellm) [reply core/mock_test.go:191]: Yes, I think we should add it but I was planning on doing it as we progress with this. So it's pending

**Comment by b00ris** [comment]
The solution looks good to me.

File: core/ibft.go [hunk 2]
```diff
@@ -453,7 +452,7 @@ func (i *IBFT) handleRoundChangeMessage(view *proto.View, quorum uint64) *proto.
 		isValidFn,
 	)
 
-	if len(msgs) < int(quorum) {
+	if i.backend.HasQuorum(view, msgs) {
```
Comment (Stefan-Ethernal) [inline core/ibft.go:455]: This needs to be negated I believe?

Comment (Stefan-Ethernal) [inline core/ibft.go:455]: I believe we need negation here:

```suggestion
	if !i.backend.HasQuorum(view, msgs) {
```

File: core/mock_test.go [hunk 3]
```diff
@@ -181,6 +181,23 @@ func (m mockBackend) BuildRoundChangeMessage(
...
+	quorum := m.Quorum(view.Height)
+
+	if len(messages) > 0 {
+		switch messages[0].GetType() {
+		case proto.MessageType_PREPREPARE:
+			return len(messages) <= 1
+		case proto.MessageType_PREPARE:
+			return len(messages) < int(quorum)-1
+		case proto.MessageType_ROUND_CHANGE, proto.MessageType_COMMIT:
+			return len(messages) < int(quorum)
```
Comment (Stefan-Ethernal) [inline core/mock_test.go:194]: Strange that function is named like this but instead is checking whether we haven't reached quorum.

Therefore suggesting changing logical conditions as following:

```suggestion
		switch messages[0].GetType() {
		case proto.MessageType_PREPREPARE:
			return len(messages) > 1
		case proto.MessageType_PREPARE:
			return len(messages) >= int(quorum)-1
		case proto.MessageType_ROUND_CHANGE, proto.MessageType_COMMIT:
			return len(messages) >= int(quorum)
```

Btw, do we need all these cases here? Seems like we need only for `PREPARE` and `ROUND_CHANGE` messages if I'm correct (was looking where we are checking `HasQuorum` within the `if` statements)

File: core/ibft.go [hunk 4]
```diff
@@ -797,7 +799,7 @@ func (i *IBFT) handlePrepare(view *proto.View, quorum uint64) bool {
 		isValidPrepare,
 	)
 
-	if len(prepareMessages) < int(quorum)-1 {
+	if i.backend.HasQuorum(view, prepareMessages) {
```
Comment (Stefan-Ethernal) [inline core/ibft.go:802]: Negate this condition, after you change `HasQuorum` implementation in the `MockBackend`:

```suggestion
	if !i.backend.HasQuorum(view, prepareMessages) {
```

**Review by Stefan-Ethernal — COMMENTED** [review]
Generally looks ok at first glance, but leaving some comments. Those are targeted to inverting logical conditions on two places, since function is called `HasQuorum` and logic is doing opposite to what it is supposed to do.

Will need to do more deep dive into `event_subscription` and `event_manager`, because that part is still abstract to me.

File: core/ibft.go (hunk 2 above)
Comment (vcastellm) [reply core/ibft.go:455]: yup

File: core/mock_test.go (hunk 3 above)
Comment (vcastellm) [reply core/mock_test.go:194]: You're right conditions are reversed, thx! 

The other extra cases are for subscriptions they also depend on those extra message types.

File: core/mock_test.go [hunk 5]
```diff
@@ -181,6 +181,23 @@ func (m mockBackend) BuildRoundChangeMessage(
 	}
 }
 
+func (m mockBackend) HasQuorum(view *proto.View, messages []*proto.Message) bool {
```
Comment (JekaMas) [inline core/mock_test.go:184]: Could we use height and messages int instead of view and *proto.messages?

Comment (vcastellm) [reply core/mock_test.go:184]: Good one, I'd say yes for the current implementation, but I guess we'll need the `From` field from the `Message` to compute voting power.

Comment (JekaMas) [reply core/mock_test.go:184]: Could we then have 1 narrow interfaces instead and then extend them is needed?

Comment (vcastellm) [reply core/mock_test.go:184]: You mean encapsulating params in an interface? can you give an example?

**Comment by vcastellm** [comment]
Closing in favor of https://github.com/0xPolygon/go-ibft/pull/39

**Pull request closed without merging**